django/db/models/query_utils.py

```diff
@@ -141,7 +141,7 @@ def __get__(self, instance, cls=None):
             return self
         data = instance.__dict__
         field_name = self.field.attname
-        if data.get(field_name, self) is self:
+        if field_name not in data:
```

Comment:
These conditions are not equivalent. I believe that Malcolm added this intentionally in dded5f52cc1dbe0e58b64a3217e12e72da7bed23. Can you find a commit that made it unnecessary?

Reply:
I _think_ this is an artifact of the past.

From [Python 2.4.1 release notes](https://www.python.org/download/releases/2.4.1/notes/)

> Dictionary objects now support the "in" operator: "x in dict" means the same as dict.has_key(x).

Given [deferred fields were introduced in Django 1.1](https://docs.djangoproject.com/en/3.0/releases/1.1/#deferred-fields), that [it was the last version to support Python 2.3](https://code.djangoproject.com/ticket/8943#comment:2) and [`dict.get` existed in Python 2.3](https://docs.python.org/release/2.3/lib/typesmapping.html) I suspect this compatibility shim has simply been carried over through releases since then.

Reply:
In [the original patch](https://code.djangoproject.com/attachment/ticket/10695/defer.6.diff) provided by Alex Gaynor we have:
```python
if self.field_name not in instance.__dict__:
```
but Malcolm changed this line to the
```python
if data.get(self.field_name, self) is self:
```
That's why I assumed that it was possible that `instance.__dict__` contained a `DeferredAttribute` instance. :thinking:

Reply:
Well `self.field_name not in instance.__dict__` is not Python 2.3 compatible so I assume it was the reason 🤷‍♂ 

As of things are right now it's not possible for `DeferredAttribute` to end up in `instance.__dict__`.

Reply:
@charettes Thanks :+1: